### PR TITLE
Automatically adjust height of iframe based on the number of choices.

### PR DESCRIPTION
**Pull request by rupalkhilari**
1) Adds an iframe Preview in embed_instructions.html, uses Javascript to dynamically fix the height of the iframe to be embedded based on the height of its content. The width however is set to the max-width of the body (350px) since the footer always seems to require it.
2) Made textarea readonly, and disabled spellcheck.

https://github.com/electology/approval_frame/issues/34

### CODE REVIEW

**Comment by ashimaathri** [comment]
@rupalkhilari I ran the server with your changes and it looks good. I really liked the idea of a preview. I will look through the PR and add my comments.

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 1]
```diff
@@ -1,7 +1,43 @@
 {% extends 'base.html' %}
 
 {% block content %}
-  Put this code in your site:
-  <textarea rows=3 cols=100><iframe src='{{ link }}' width='250'></iframe></textarea>
+
+<script type='text/javascript'>
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: It's best to separate the JavaScript code from the HTML template by putting it in the [static](https://github.com/electology/approval_frame/tree/master/approval_polls/static/approval_polls) directory. You can then include the JavaScript file using the `staticfiles` app similar to how `create.js` is included [here](https://github.com/electology/approval_frame/blob/master/approval_polls/templates/approval_polls/create.html#L7)

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 2]
```diff
@@ -1,7 +1,43 @@
...
+    // Since the width of the embedded document's body is restricted to 350px, and displaying
+    // the footer would always require the max-width, setting the iframe's width to 350px.
+    textArea.value = "<iframe src='{{ link }}' width='350px' height='"+dynHeight+"'></iframe>";
+}
+
+</script>
+<p>Put this code in your site:<p>
+<textarea rows=3 cols=100 id='textAreaCode' spellcheck='false' readonly></textarea>
+<p>Preview:<p>
+<iframe id="iframePreview" name="iframePreview" src='{{ link }}' onload="setIframeHeightAndCode(this.id)" width='350px'>Sorry, your browser doesn't support iframes.</iframe>
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: Instead of calling JavaScript in the HTML using the onload callback, it is better to get the element by id in the JavaScript code and manipulate the properties of it there.

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 3]
```diff
@@ -1,7 +1,43 @@
...
-  Put this code in your site:
-  <textarea rows=3 cols=100><iframe src='{{ link }}' width='250'></iframe></textarea>
+
+<script type='text/javascript'>
+
+// thanks to http://www.dyn-web.com
+function getDocHeight(doc) {
+    doc = doc || document;
+    var body = doc.body, html = doc.documentElement;
+    var height = Math.max( body.scrollHeight, body.offsetHeight,
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: It will be better to use libraries like jQuery/Dojo/Google Closure to do DOM manipulation (like setting/getting properties of nodes) as these libraries are tried and tested and support most popular browsers. This way, the code will also be more concise.
For example to get the height of the body using jQuery would just be $('body').height()

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 4]
```diff
@@ -1,7 +1,43 @@
...
 
 {% block content %}
-  Put this code in your site:
-  <textarea rows=3 cols=100><iframe src='{{ link }}' width='250'></iframe></textarea>
+
+<script type='text/javascript'>
+
+// thanks to http://www.dyn-web.com
+function getDocHeight(doc) {
+    doc = doc || document;
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: Should be `var doc` to avoid global namespace pollution

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 5]
```diff
@@ -1,7 +1,43 @@
...
+// thanks to http://www.dyn-web.com
+function setIframeHeightAndCode(id) {
+    var ifrm = document.getElementById(id);
+    var doc = ifrm.contentDocument? ifrm.contentDocument: ifrm.contentWindow.document;
+    ifrm.style.visibility = 'hidden';
+    ifrm.style.height = "10px"; 
+    ifrm.style.height = getDocHeight( doc ) + "px";
+    ifrm.style.visibility = 'visible';
+    // Write out the required HTML iframe code containing this dynamic height into the textarea.
+    setIframeDrawCode('textAreaCode', ifrm.style.height);
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: Same comment as above (to use a library)

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 6]
```diff
@@ -1,7 +1,43 @@
...
+        html.clientHeight, html.scrollHeight, html.offsetHeight );
+    return height;
+}
+
+// thanks to http://www.dyn-web.com
+function setIframeHeightAndCode(id) {
+    var ifrm = document.getElementById(id);
+    var doc = ifrm.contentDocument? ifrm.contentDocument: ifrm.contentWindow.document;
+    ifrm.style.visibility = 'hidden';
+    ifrm.style.height = "10px";
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: For consistency, we'll be using single quotes for strings everywhere possible.

**Comment by ashimaathri** [comment]
@rupalkhilari I've added jQuery as a dependency. It is included in the head section of [base.html](https://github.com/electology/approval_frame/blob/master/approval_polls/templates/base.html#L6) and [public_base.html](https://github.com/electology/approval_frame/blob/master/approval_polls/templates/public_base.html#L5) so you can start using jQuery from within any script that is included after it.

**Comment by rupalkhilari** [comment]
Thanks for the comments, Ashima. I will look to implement this with jQuery in a separate file.

File: approval_polls/templates/approval_polls/embed_instructions.html (hunk 2 above)
Comment (rupalkhilari) [inline approval_polls/templates/approval_polls/embed_instructions.html]: In the latest commit (using JQuery), I have called the required code from $(window).load(function()) since the contents size could only be determined after it was loaded. 
Hope that is alright.

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+
+$(window).load(function() {
+
+    var iframeId = '#iframePreview';
+    var textareaId = '#textAreaCode';
+    var pollLink = $(iframeId).attr('src');
+
+    // Set the iframe height and textarea code.
+    setIframeHeightAndCode(iframeId, textareaId, pollLink);
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: Can shorten this to `setIframeHeightAndCode('#iframePreview', '#textAreaCode', $(iframeId).attr('src'))`

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+
+$(window).load(function() {
+
+    var iframeId = '#iframePreview';
+    var textareaId = '#textAreaCode';
+    var pollLink = $(iframeId).attr('src');
+
+    // Set the iframe height and textarea code.
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: You've give quite good names to the functions :+1:  (i.e. they are self explanatory). So, we don't need this comment :smile_cat:

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/* Code related to templates/approval_polls/embed_instructions.html */
+
+// Sets the iframe height and the corresponding embed instruction code in the textarea.
+function setIframeHeightAndCode(iframeId, textareaId, pollLink) {
+
+    // get the height of the contents of the iframe.
+    var iframeHeight = $(iframeId).contents().height();
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: It's better to move all variable declarations to the top because the JavaScript interpreter will move them up anyway. See [this blog post on hoisting] (http://www.adequatelygood.com/JavaScript-Scoping-and-Hoisting.html) for a detailed explanation.

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    // set the height of the iframe to that of its contents.
+    $(iframeId).height(iframeHeight);
+
+    // write the HTML code into the textarea.
+    $(textareaId).val("<iframe src='"+pollLink+"' width='350px' height='"+iframeHeight+"px'></iframe>");
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: (Nitpick) For readability and code consistency, it looks better to have spaces around operators.
For example `"<iframe src='"+pollLink` vs `"<iframe src='" + pollLink`

File: approval_polls/templates/approval_polls/embed_instructions.html [hunk 11]
```diff
@@ -1,7 +1,14 @@
 {% extends 'base.html' %}
 
 {% block content %}
-  Put this code in your site:
-  <textarea rows=3 cols=100><iframe src='{{ link }}' width='250'></iframe></textarea>
+
+    {% load staticfiles %}
```
Comment (ashimaathri) [inline approval_polls/templates/approval_polls/embed_instructions.html]: This should go in the {% block head %} section instead of the content section

File: approval_polls/static/approval_polls/embed_instructions.js (hunk 9 above)
Comment (rupalkhilari) [inline approval_polls/static/approval_polls/embed_instructions.js]: A very nice explanation. I am quite new to web development and hadn't come across this before. Thanks!

**Comment by ashimaathri** [comment]
@rupalkhilari Thanks for taking the time to incorporate the feedback! The code looks good. If you could just remove the comments, I can merge it with master.

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  // obtain the src attribute of the iframe.
+  pollLink = $(iframeId).attr('src');
+
+  // write the HTML code into the textarea.
+  $(textareaId).val("<iframe src='" + pollLink + "' width='350px' height='" + iframeHeight + "px'></iframe>");
+}
+
+
+$(window).load(function() {
+
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: Remove newline

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
+/* Code related to templates/approval_polls/embed_instructions.html */
+
+// Sets the iframe height and the corresponding embed instruction code in the textarea.
+function setIframeHeightAndCode(iframeId, textareaId) {
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: Should be `setIframeHeightAndCode = function (iframeId, textareaId)` for code consistency and to make hoisting explicit.

File: approval_polls/static/approval_polls/embed_instructions.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
+/* Code related to templates/approval_polls/embed_instructions.html */
```
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: Wrap the entire file in an [IIFE](http://en.wikipedia.org/wiki/Immediately-invoked_function_expression)

File: approval_polls/static/approval_polls/embed_instructions.js (hunk 9 above)
Comment (ashimaathri) [inline approval_polls/static/approval_polls/embed_instructions.js]: Yup. I'm happy to share. Feel free to post any interesting articles you find on our google group. I too learnt JS from blogs and other people's code :smile: and I'm still learning.

**Comment by rupalkhilari** [comment]
@ashimaathri , thanks for the feedback and the insights. I have retained just one of the comments for the function. Hope it is satisfactory. 

Thanks!

**Comment by ashimaathri** [comment]
Thanks @rupalkhilari. The code was excellent and I saw that you'd followed the 80 char code style too! I will try to make a style guide for this project so I don't annoy other enthusiastic contributors like you :smile:

**Comment by rupalkhilari** [comment]
@ashimaathri , a style guide for the project would be perfect ! I just noticed https://github.com/styleguide/javascript apparently they recommend using CoffeeLint

**Comment by ashimaathri** [comment]
@rupalkhilari Yup, we could add a Grunt task for linting JS. The github guide seems to be for coffeescript though.
